massa-execution-worker/src/speculative_roll_state.rs

```diff
@@ -148,6 +149,63 @@ impl SpeculativeRollState {
         Ok(())
     }
 
+    /// Try to slash `roll_count` rolls from the given address. If not enough roll, slash
+    /// the available amount and return the value.
+    ///
+    /// # Arguments
+    /// * `addr`: address to slash the rolls from
+    /// * `roll_count`: number of rolls to slash
+    pub fn try_slash_rolls(
+        &mut self,
+        addr: &Address,
+        roll_count: u64,
+    ) -> Result<u64, ExecutionError> {
+        // fetch the roll count
+        let owned_count = self.get_rolls(addr);
+        let roll_to_slash = min(roll_count, owned_count);
+
+        // remove the rolls
+        let current_rolls = self
+            .added_changes
+            .roll_changes
+            .get_mut(addr)
+            .ok_or_else(|| {
+                ExecutionError::RuntimeError(format!("Cannot get roll changes for addr: {}", addr))
+            })?;
+
+        *current_rolls = current_rolls.saturating_sub(roll_to_slash);
+        Ok(roll_to_slash)
+    }
+
+    /// Try to slash `amount` credits from the given address. If not enough credits, slash
+    /// the available amount and return the value.
+    ///
+    /// # Arguments
+    /// * `addr`: address to slash the deferred credits from
+    /// * `amount`: number of deferred credits to slash
+    pub fn try_slash_deferred_credits(
+        &mut self,
+        slot: &Slot,
+        addr: &Address,
+        amount: &Amount,
+    ) -> Result<Amount, ExecutionError> {
+        let credits = self.get_address_deferred_credits(addr, *slot);
+
+        // Return an error here - this should never happen
```

Comment:
if there are not enough deferred credits in "last_key_value", other deferred credits (not just the ones targeted to the latest slot) should be slashed.

Ideally we would do the following:
* loop over deferred credits by increasing slot S
  * try slashing the amount from the deferred credits targeting S
    * if there was enough to slash all the coins we needed to slash, break

Reply:
Think I've fixed it :)